### PR TITLE
4.x: Workspace Refactoring: Read Workspace Artifacts

### PR DESCRIPTION
**Pull request by deepakrkris**
4.x: Workspace Refactoring: read workspace artifacts

### CODE REVIEW

File: component/datamodel/util/file-utility.js [hunk 1]
```diff
@@ -85,3 +85,24 @@ function writePackageDefinition(packageDef, cb) {
   const data = packageDef.getDefinition();
   fs.writeJson(filePath, data, cb);
 }
+
+function readModelConfig(facet, cb) {
+  const filePath = facet.getModelConfigPath();
+  fs.readJson(filePath, function(err, data) {
+    if (err) return err;
+    facet.setModelConfig(data);
```
Comment (superkhau) [inline component/datamodel/util/file-utility.js]: cb()?

File: component/datamodel/util/file-utility.js [hunk 2]
```diff
@@ -85,3 +85,24 @@ function writePackageDefinition(packageDef, cb) {
...
+function readModel(id, workspace, cb) {
+  const filePath = workspace.getModelDefinitionPath(id);
+  fs.readJson(filePath, cb);
+}
+
+function readDataSourceConfig(workspace, cb) {
+  const filePath = workspace.getDataSourceConfigFilePath();
+  fs.writeJson(filePath, function(err, data) {
+    if (err) return err;
+    workspace.setDataSources(data);
```
Comment (superkhau) [inline component/datamodel/util/file-utility.js]: cb()?

File: component/workspace.js [hunk 3]
```diff
@@ -56,6 +57,15 @@ class Workspace extends Graph {
...
     return facet;
   }
+  createModelDefinition(id, modelDef) {
+    const modelData = clone(modelDef);
+    delete modelData['properties'];
+    delete modelData['methods'];
+    delete modelData['relations'];
+    delete modelData['validations'];
+    delete modelData['acls'];
+    new Model(this, id, modelData);
```
Comment (superkhau) [inline component/workspace.js]: add comment regarding side effect of `new Model`

File: connector/connector.js [hunk 4]
```diff
@@ -70,3 +70,35 @@ connector.createMiddleware = function(phase, data, cb) {
   middlewareDef,
   cb);
 };
+
+connector.createModel = function(id, data, cb) {
```
Comment (superkhau) [inline connector/connector.js]: rename this file to `index.js`

Comment (superkhau) [inline connector/connector.js]: Something weird about the indentation and contents of this file, is there a module.exports I'm missing?

File: connector/model-handler.js [hunk 5]
```diff
@@ -27,5 +27,37 @@ class ModelHandler {
     const taskList = [create];
     workspace.execute(taskList, callback);
   }
+
+  static findModelConfig(workspace, modelId, cb) {
+    const taskList = [];
+    workspace.getFacets().forEach(function(facet) {
+      taskList.push(function(next) {
+        workspace.refreshModelConfig(facet.getName(), function(err) {
```
Comment (superkhau) [inline connector/model-handler.js]: workspace.refreshModelConfig(facet.getName(), next);

File: connector/model-handler.js [hunk 6]
```diff
@@ -27,5 +27,37 @@ class ModelHandler {
...
+      if (err) return cb(err);
+      const modelConfig = workspace.getModelConfig(modelId);
+      cb(null, modelConfig.getDefinition());
+    }
+    workspace.execute(taskList, callback);
+  }
+
+  static findModel(workspace, modelId, cb) {
+    function refresh(next) {
+      workspace.refreshModel(modelId, function(err) {
```
Comment (superkhau) [inline connector/model-handler.js]: workspace.refreshModel(modelId, next);

File: connector/connector.js (hunk 4 above)
Comment (deepakrkris) [inline connector/connector.js]: agreed , changed to index.js. here the memory connector is accessed via app.datasources, so there is no module.exports. we are not creating, we are customizing an existing object.

File: connector/model-handler.js (hunk 5 above)
Comment (deepakrkris) [inline connector/model-handler.js]: agreed

File: connector/model-handler.js (hunk 6 above)
Comment (deepakrkris) [inline connector/model-handler.js]: agreed

File: component/workspace.js (hunk 3 above)
Comment (deepakrkris) [inline component/workspace.js]: agreed

File: component/datamodel/util/file-utility.js (hunk 1 above)
Comment (deepakrkris) [inline component/datamodel/util/file-utility.js]: agreed

File: common/models/data-source-definition.js [hunk 7]
```diff
@@ -18,8 +18,17 @@ module.exports = function(DataSourceDefinition) {
       }
       const connector = DataSourceDefinition.getConnector();
       const id = data.id;
-      //TODO(Deepak) - add response handling later as part of the callback
+      // TODO(Deepak) - add response handling later as part of the callback
```
Comment (superkhau) [inline common/models/data-source-definition.js]: I still think you should leave the TODO notes outside the source code and in your own notes. Just delete the lines altogether and fix it when the time comes.

File: common/models/middleware.js [hunk 8]
```diff
@@ -24,7 +24,7 @@ module.exports = function(Middleware) {
       }
       const phase = this.getPhase(data);
       const connector = Middleware.getConnector();
-      //TODO(Deepak) - add response handling later
+      // TODO(Deepak) - add response handling later
```
Comment (superkhau) [inline common/models/middleware.js]: ditto

File: common/models/model-config.js [hunk 9]
```diff
@@ -19,8 +19,18 @@ module.exports = function(ModelConfig) {
       }
       const id = data.id;
       const connector = ModelConfig.getConnector();
-      //TODO(Deepak) - add response handling later
+      // TODO(Deepak) - add response handling later
```
Comment (superkhau) [inline common/models/model-config.js]: ditto

File: common/models/model-config.js [hunk 10]
```diff
@@ -19,8 +19,18 @@ module.exports = function(ModelConfig) {
...
       connector.createModelConfig(id, data, cb);
     };
+    ModelConfig.find = function(filter, options, cb) {
+      if (typeof options === 'function') {
+        cb = options;
+        options = null;
+      }
+      const id = filter.where.id;
+      const connector = ModelConfig.getConnector();
+      // TODO(Deepak) - add response handling later
```
Comment (superkhau) [inline common/models/model-config.js]: ditto

File: common/models/model-definition.js [hunk 11]
```diff
@@ -18,8 +18,17 @@ module.exports = function(ModelDefinition) {
         options = null;
       }
       const id = data.id;
-      //TODO(Deepak) - add response handling later
+      // TODO(Deepak) - add response handling later
```
Comment (superkhau) [inline common/models/model-definition.js]: ditto

File: common/models/model-definition.js [hunk 12]
```diff
@@ -18,8 +18,17 @@ module.exports = function(ModelDefinition) {
...
       connector.createModel(id, data, cb);
     };
+    ModelDefinition.find = function(filter, options, cb) {
+      if (typeof options === 'function') {
+        cb = options;
+        options = null;
+      }
+      const id = filter.where.id;
+      // TODO(Deepak) - add response handling later
```
Comment (superkhau) [inline common/models/model-definition.js]: ditto

File: component/datamodel/facet.js [hunk 13]
```diff
@@ -30,19 +33,24 @@ class Facet extends Node {
...
-      }
-    });
-    return config;
+    if (modelConfigNodes) {
+      Object.keys(modelConfigNodes).forEach(function(key) {
+        config[key] = modelConfigNodes[key]._content;
+      });
+    }
+    if (id) return config[id];
+    else return config;
```
Comment (superkhau) [inline component/datamodel/facet.js]: `return id ? config[id] : config;`

File: component/datamodel/model.js [hunk 14]
```diff
@@ -52,6 +52,15 @@ class Model extends Node {
     modelDef.relations = relations;
     return modelDef;
   }
+  updateDefinition(modelDef) {
+    var modelData = clone(modelDef);
+    delete modelData['properties'];
+    delete modelData['methods'];
+    delete modelData['relations'];
+    delete modelData['validations'];
+    delete modelData['acls'];
```
Comment (superkhau) [inline component/datamodel/model.js]: maybe lexico sort these?

File: component/datamodel/util/file-utility.js [hunk 15]
```diff
@@ -9,6 +9,9 @@ const fs = require('fs-extra');
 const path = require('path');
 
 module.exports = {
+  readModel: readModel,
+  readModelConfig: readModelConfig,
+  readDataSource: readDataSource,
```
Comment (superkhau) [inline component/datamodel/util/file-utility.js]: lexico: readDataSource should be the first item in the list

File: component/datamodel/util/file-utility.js [hunk 16]
```diff
@@ -85,3 +88,26 @@ function writePackageDefinition(packageDef, cb) {
   const data = packageDef.getDefinition();
   fs.writeJson(filePath, data, cb);
 }
+
+function readModelConfig(facet, cb) {
+  const filePath = facet.getModelConfigPath();
+  fs.readJson(filePath, function(err, data) {
+    if (err) return err;
```
Comment (superkhau) [inline component/datamodel/util/file-utility.js]: if (err) return cb(err);

File: component/tasks.js [hunk 17]
```diff
@@ -50,17 +50,44 @@ class Tasks {
...
+    fsUtility.readModel(facetName, modelName, workspace,
+    function(err, modelDef) {
+      if (err) return cb(err);
+      if (model) {
+        model.updateDefinition(modelDef);
+        cb(null, model.getDefinition());
+      } else {
+        workspace.createModelDefinition(modelId, modelDef);
+        cb(null, model.getDefinition());
+      }
```
Comment (superkhau) [inline component/tasks.js]:
```
if (model) {
  model.updateDefinition(modelDef);
} else {
  workspace.createModelDefinition(modelId, modelDef);
}
cb(null, model.getDefinition());

File: component/workspace.js [hunk 18]
```diff
@@ -52,10 +53,28 @@ class Workspace extends Graph {
...
+      config.ModelDefaultDir, modelName + '.json');
+    return filePath;
+  }
+  createModelDefinition(id, modelDef) {
+    const modelData = clone(modelDef);
+    delete modelData['properties'];
+    delete modelData['methods'];
+    delete modelData['relations'];
+    delete modelData['validations'];
+    delete modelData['acls'];
```
Comment (superkhau) [inline component/workspace.js]: alphabetical?

File: connector/data-source-handler.js [hunk 19]
```diff
@@ -13,5 +13,20 @@ class DataSourceHandler {
     const taskList = [create];
     workspace.execute(taskList, callBack);
   }
+
+  static findDataSource(workspace, id, cb) {
+    function refresh(next) {
+      workspace.refreshDataSource(function(err) {
+        next(err);
+      });
```
Comment (superkhau) [inline connector/data-source-handler.js]: workspace.refreshDataSource(next);

File: connector/data-source-handler.js [hunk 20]
```diff
@@ -13,5 +13,20 @@ class DataSourceHandler {
...
+    };
+    function callBack(err, results) {
```
Comment (superkhau) [inline connector/data-source-handler.js]: callback with lowercase B

File: connector/model-handler.js [hunk 21]
```diff
@@ -27,5 +27,41 @@ class ModelHandler {
...
     workspace.execute(taskList, callback);
   }
+
+  static findModelConfig(workspace, modelId, cb) {
+    const taskList = [];
+    const facets = workspace.getFacets();
+    Object.keys(facets).forEach(function(key) {
+      let facet = facets[key];
+      taskList.push(function(next) {
+        workspace.refreshModelConfig(facet.getName(), function(err) {
```
Comment (superkhau) [inline connector/model-handler.js]: workspace.refreshModelConfig(facet.getName(), next);

File: test/acceptance-tests/use-cases/test-scripts/create-datasource.js [hunk 22]
```diff
@@ -41,4 +41,23 @@ module.exports = function() {
...
+    DataSourceDefinition.find(filter, function(err, data) {
+      if (err) return next(err);
+      testsuite.datasource = data;
+      next();
+    });
+  });
+
+  this.Then(/^the datasource definition is returned$/, function(next) {
+    expect(Object.keys(testsuite.datasource))
+    .to.include.members(['connector', 'name']);
```
Comment (superkhau) [inline test/acceptance-tests/use-cases/test-scripts/create-datasource.js]: indent one level

File: test/acceptance-tests/use-cases/test-scripts/find-models.js [hunk 23]
```diff
@@ -0,0 +1,60 @@
...
+const fs = require('fs-extra');
+const loopback = require('loopback');
+const ModelClass = require('../../../../component/datamodel/model');
+const path = require('path');
+const testSupport = require('../../../helpers/test-support');
+const util = require('util');
+const workspaceManager = require('../../../../component/workspace-manager');
+
+const ModelDefinition = app.models.ModelDefinition;
+const ModelConfig = app.models.ModelConfig;
```
Comment (superkhau) [inline test/acceptance-tests/use-cases/test-scripts/find-models.js]: Move these into the declarations inside the module.exports (group with `testsuite`)

File: test/acceptance-tests/use-cases/test-scripts/find-models.js [hunk 24]
```diff
@@ -0,0 +1,60 @@
...
+
+module.exports = function() {
+  var testsuite = this;
```
Comment (superkhau) [inline test/acceptance-tests/use-cases/test-scripts/find-models.js]: let testsuite = this;

File: connector/connector.js (hunk 4 above)
Comment (superkhau) [inline connector/connector.js]: I see, why don't we rename this to workspace-connector? It looks like you're basically extending connector (well mixing in new behaviours).

**Comment by deepakrkris** [comment]
closed in favor of #389 #388 #387

**Pull request closed without merging**